### PR TITLE
aws_cloudtrail support

### LINKED ISSUE #94: Support for CloudTrail resources
Requesting addition of AWS CloudTrail trails as time permits.

### PR DESCRIPTION
**Pull request by aghassemlouei**
**Description:** 
Adding support for aws cloudtrail trail deletion

**Justification:** 
1. Expand support for service coverage 
2. Ensure s3 buckets can be deleted when cloudtrail is logging to s3 
3. Resolves #94

### CODE REVIEW

**Comment by aghassemlouei** [comment]
Apologies for the incredibly poor variable names but it's not entirely clear what the convention is and I'm 100% down for any modifications. Attempted to follow instructions on #21 but it seems like it's missing a step or two but I was able to kind of follow along by referencing a few other pull requests that had made their way to master. Might be worth converting the issue into a contributing guide?

Unfortunately I still managed to miss something with the travis build output but I think it may be unrelated as my other seemingly innocuous pull request #95 also seemed to fail? Unfortunately the integration tests are beyond my capabilities and bandwidth but I promise I'll give it a shot one day!

File: all.yml [hunk 1]
```diff
@@ -1,6 +1,7 @@
 aws_ami:
 aws_autoscaling_group:
 aws_cloudformation_stack:
+aws_cloudtrail_trail:
```
Comment (jckuester) [inline all.yml]: FYI, we use Terraform resource types as a resource type identifier. I assume that you mean the `aws_cloudtrail` resource? See: https://www.terraform.io/docs/providers/aws/r/cloudtrail.html

File: resource/supported.go [hunk 2]
```diff
@@ -73,6 +75,7 @@ const (
 	S3Bucket            TerraformResourceType = "aws_s3_bucket"
 	SecurityGroup       TerraformResourceType = "aws_security_group"
 	Subnet              TerraformResourceType = "aws_subnet"
+	Trail               TerraformResourceType = "aws_cloudtrail_trail"
```
Comment (jckuester) [inline resource/supported.go]:
```suggestion
	CloudTrail               TerraformResourceType = "aws_cloudtrail"
```

File: all.yml (hunk 1 above)
Comment (jckuester) [inline all.yml]:
```suggestion
aws_cloudtrail:
```

File: README.md [hunk 3]
```diff
@@ -168,6 +168,7 @@ A technical reason for this is that AWSweeper is build upon the already existing
 | aws_ami                          | x             | x
 | aws_autoscaling_group            | x             | x
 | aws_cloudformation_stack         | x             | x
+| aws_cloudtrail_trail             |               |
```
Comment (jckuester) [inline README.md]:
```suggestion
| aws_cloudtrail             |               |
```

File: resource/supported.go [hunk 4]
```diff
@@ -309,6 +315,8 @@ func (a *AWS) RawResources(resType TerraformResourceType) (interface{}, error) {
 		return a.SecurityGroup()
 	case Subnet:
 		return a.subnets()
+	case Trail:
+		return a.trails()
```
Comment (jckuester) [inline resource/supported.go]:
```suggestion
		return a.cloudTrails()
```

Comment (jckuester) [inline resource/supported.go]: would be great if you could rename everywhere accordingly :+1:

File: resource/supported.go [hunk 5]
```diff
@@ -111,6 +114,7 @@ var (
 		S3Bucket:            "Name",
 		SecurityGroup:       "GroupId",
 		Subnet:              "SubnetId",
+		Trail:               "TrailName",
```
Comment (jckuester) [inline resource/supported.go]:
```suggestion
		Trail:               "Name",
```

File: resource/supported.go [hunk 6]
```diff
@@ -595,6 +603,13 @@ func (a *AWS) s3Buckets() (interface{}, error) {
 	return output.Buckets, nil
 }
 
+func (a *AWS) trails() (interface{}, error) {
+	output, err := a.ListTrails(&cloudtrail.ListTrailsInput{})
```
Comment (jckuester) [inline resource/supported.go]: It's not obvious, but I think we should use the `DescribeTrails` function here instead.

**Comment by jckuester** [comment]
Hi @aghassemlouei :wave: 

Thanks for the PR, I gave you some feedback. I can take care of the integration test once you are done.

A contribution guide is probably a good idea. However, I can give you already a sneak peek into my plans of auto-generating the support for resources (https://github.com/jckuester/awsls). I auto-generated already list functions for over 200 resources and will integrate the feature soon into awsweeper. Stay tuned :smiley:

File: all.yml (hunk 1 above)
Comment (aghassemlouei) [reply all.yml]: Got it; apologies for having to hand-hold me through this pull request.

Comment (aghassemlouei) [reply all.yml]: Done and done; thank you for the pointer!

File: README.md (hunk 3 above)
Comment (aghassemlouei) [reply README.md]: Addressed!

File: resource/supported.go (hunk 5 above)
Comment (aghassemlouei) [reply resource/supported.go]: Adjusted; I am slowly understanding what needs to get modified and why.

File: resource/supported.go (hunk 2 above)
Comment (aghassemlouei) [reply resource/supported.go]: Adjusted but why did you derive `CloudTrail` from? Technically it's a service and within it you can have multiple trails but yeah, no big deal.

File: resource/supported.go (hunk 4 above)
Comment (aghassemlouei) [reply resource/supported.go]: Yep yep; apologies for this i saw inconsistencies with things like KmsAliases and got turned around quite a bit.

Comment (aghassemlouei) [reply resource/supported.go]: Adjusted!

File: resource/supported.go (hunk 6 above)
Comment (aghassemlouei) [reply resource/supported.go]: Done and done! Yeah, always imagined a `describe` would always end up with us having to manipulate or parse the output. Any reason behind the switch from your perspective?

**Comment by aghassemlouei** [comment]
Thank you for taking the time to show and suggest the modifications. If you could integrate the `awsls` utility to feed this and near instantly expand the service coverage that'd be amazing!

File: resource/supported.go (hunk 2 above)
Comment (jckuester) [reply resource/supported.go]: Correct, it's a service, but in Terraform each trail is defined by a resource type called `aws_cloudtrail` (see here: https://www.terraform.io/docs/providers/aws/r/cloudtrail.html)

**Comment by jckuester** [comment]
Thanks @aghassemlouei for the quick turn-around on this PR. Merged!

**Comment by jckuester** [comment]
FYI, here are the tests I created for this PR: https://github.com/cloudetc/awsweeper/pull/99

**Pull request merged**